arch/arm/src/sama5/sam_classd.c

```diff
@@ -0,0 +1,2617 @@
+      case CLASSD_DSP_FREQ_11M2896:
+        {
+          regval = PMC_AUDIO_PLL0_ND(DSP_CK_ND_11M1896) |
+                   PMC_AUDIO_PLL0_QDPMC(DSP_CK_QDPMC_11M1896) |
+                   PMC_AUDIO_PLL0_FLT(PMC_AUDIO_PLL0_PLLFLT_STD) |
+                   PMC_AUDIO_PLL0_RESETN;
+          classd_putreg(SAM_PMC_AUDIO_PLL0, regval);
+
+          regval = PMC_AUDIO_PLL1_FRACR(DSP_CK_FRACR_11M1896)  |
+                   PMC_AUDIO_PLL1_DIV(DSP_CK_DIV_11M1896)      |
+                   PMC_AUDIO_PLL1_QDAUDIO(DSP_QDAUDIO_11M1896) |
+                   PMC_AUDIO_PLL1_DIV(2);
+          classd_putreg(SAM_PMC_AUDIO_PLL1, regval);
+        }
+        break;
+      default: /* CLASSD_DSP_FREQ_12M288 */
+        {
+          regval = PMC_AUDIO_PLL0_ND(DSP_CK_ND_12M288) |
+                   PMC_AUDIO_PLL0_QDPMC(DSP_CK_QDPMC_12M288) |
+                   PMC_AUDIO_PLL0_FLT(PMC_AUDIO_PLL0_PLLFLT_STD) |
+                   PMC_AUDIO_PLL0_RESETN;
+          classd_putreg(SAM_PMC_AUDIO_PLL0, regval);
+
+          regval = PMC_AUDIO_PLL1_FRACR(DSP_CK_FRACR_12M288)  |
+                   PMC_AUDIO_PLL1_DIV(DSP_CK_DIV_12M288)      |
+                   PMC_AUDIO_PLL1_QDAUDIO(DSP_QDAUDIO_12M288) |
+                   PMC_AUDIO_PLL1_DIV(2);
+          classd_putreg(SAM_PMC_AUDIO_PLL1, regval);
+        }
+        break;
+    }
+#endif
+
+  classd_dump_registers("After dsp clock setup");
+
+#ifdef CONFIG_CLASSD_REGDEBUG
+  nd = (pll0 & PMC_AUDIO_PLL0_ND_MASK) >> PMC_AUDIO_PLL0_ND_SHIFT;
+  fracr = (pll1 & PMC_AUDIO_PLL1_FRACR_MASK) >> PMC_AUDIO_PLL1_FRACR_SHIFT;
+  qdpmc = (pll0 & PMC_AUDIO_PLL0_QDPMC_MASK) >> PMC_AUDIO_PLL0_QDPMC_SHIFT;
+
+  clk *= ((nd + 1) << 22) + fracr;
+  clk /= 1 << 22;
+  clk /= (qdpmc + 1);
+  clk /= 8;
+
+  audinfo("INFO: audio pll clk = %" PRId64 "\n", clk);
+#endif
+}
+
+/****************************************************************************
+ * Name: classd_setup
+ *
+ * Description:
+ *   Configure the ClassD peripheral.
+ *
+ * Input Parameters:
+ *   priv - A reference to the CLASSD peripheral state.
+ *
+ * Returned Value:
+ * OK (or not)
+ *
+ ****************************************************************************/
+
+static int classd_setup(struct classd_dev_s *priv)
+{
+  uint32_t regval;
+
+  classd_set_mode(priv);
+  classd_set_interpolator(priv);
+  classd_dsp_clock_config(priv);
+
+  regval = classd_getreg(priv, SAM_CLASSD_INTSR);
+  if (regval != 0)
+    {
+      return -EINVAL;
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: classd_getreg
+ *
+ * Description:
+ *   Read the value of a CLASSD register.
+ *
+ * Input Parameters:
+ *   priv - A reference to the CLASSD peripheral state
+ *   offset - The offset to the register to read
+ *
+ * Returned Value:
+ *   Value of the register requested
+ *
+ ****************************************************************************/
+
+uint32_t classd_getreg(struct classd_dev_s *priv, uint32_t regaddr)
```

Comment:
Is it used outside of this file? I mean, should it be static?

Reply:
Yes, it should be static.

Reply:
Let's add it.

Reply:
I have - and will commit once the remaining issues are resolved.

Reply:
done